src/io/sparse_bin.hpp

```diff
@@ -92,7 +92,9 @@ class SparseBin : public Bin {
   void Push(int tid, data_size_t idx, uint32_t value) override {
     auto cur_bin = static_cast<VAL_T>(value);
     if (cur_bin != 0) {
-      push_buffers_[tid].emplace_back(idx, cur_bin);
+      // Splitting this operation into separate calls avoids a concurrency exception. Do not simplify.
+      auto buffer = push_buffers_[tid];
```

Comment:
this will be a clone, rather than a reference?

Reply:
it should be the same either way (reference), but for some reason splitting it into 2 calls stopped the exceptions.  I will keep investigating to see if I can determine why.

The main point of this PR was to add in the multi-threading tests, since in the previous PR I promised I would add them in a subsequent PR.

Reply:
I believe most "=" operator in C++ is "copy assignment", refer to https://www.enseignement.polytechnique.fr/informatique/INF478/docs/Cpp/en/cpp/container/vector/operator%3D.html 
I also write a small test.
```
#include <iostream>
#include <vector>
int main() {
    std::vector<int> a = {1, 2, 3};
    auto b = a;
    printf("%d, %d\n", &a, &b);
    return 0;
}
```
And it shows the addresses of `a` and `b` is different. 

And you should use `auto& b = a` to get the reference.

Reply:
yeah, I removed this since it was wrong and as you said, didn't do what I thought it was doing.